main.c

```diff
@@ -819,33 +820,91 @@ resolve_all_solvables (Pool  *pool,
   return solv_failed;
 }
 
-static Queue
-pile_from_solvables (Pool  *pool,
-                     GStrv  solvables)
+static void
+add_solvable_to_pile (const char *solvable,
+                      Pool       *pool,
+                      Queue      *pile)
 {
-  Queue pile;
-  queue_init (&pile);
   int sel_flags = SELECTION_NAME | SELECTION_PROVIDES | SELECTION_GLOB |
                   SELECTION_CANON | SELECTION_DOTARCH;
   g_auto(Queue) sel;
   queue_init (&sel);
+  selection_make (pool, &sel, solvable, sel_flags);
+  g_auto(Queue) q;
+  queue_init (&q);
+  selection_solvables (pool, &sel, &q);
+  if (!q.count)
+    {
+      g_warning ("Nothing matches '%s'", solvable);
+      return;
+    }
+  pool_best_solvables (pool, &q, 0);
+  for (int j = 0; j < q.count; j++)
+    queue_push (pile, q.elements[j]);
+}
+
+static void
+add_solvables_from_file_to_pile (const char  *filename,
+                                 Pool        *pool,
+                                 Queue       *pile,
+                                 GError     **error)
+{
+  GIOChannel *ch = g_io_channel_new_file (filename, "r", error);
+  if (ch == NULL)
+    {
+      g_prefix_error (error, "Unable to open file '%s': ", filename);
+      return;
+    }
+
+  GIOStatus ret;
+  do
+    {
+      g_autofree gchar *content = NULL;
+      gsize length, tpos;
+
+      ret = g_io_channel_read_line (ch, &content, &length, &tpos, error);
+      if (ret != G_IO_STATUS_NORMAL || content == NULL)
+        continue;
+
+      content[tpos] = '\0';
+      if (*content)
+        add_solvable_to_pile (content, pool, pile);
+    }
+  while (ret == G_IO_STATUS_NORMAL);
+
+  if (ret != G_IO_STATUS_EOF)
+    {
+      g_prefix_error (error, "Failure while reading file '%s': ", filename);
+      return;
+    }
+
+  /* make sure any error is cleared to indicate success */
+  g_clear_error (error);
+}
+
+static Queue *
+pile_from_solvables (Pool    *pool,
+                     GStrv    solvables,
+                     GError **error)
+{
+  g_return_val_if_fail (*error == NULL, NULL);
+
+  g_autoptr(Queue) pile = g_malloc (sizeof (Queue));
+  queue_init (pile);
+
   for (GStrv solvable = solvables; solvable && *solvable; solvable++)
     {
-      selection_make (pool, &sel, *solvable, sel_flags);
-      g_auto(Queue) q;
-      queue_init (&q);
-      selection_solvables (pool, &sel, &q);
-      if (!q.count)
-        {
-          g_warning ("Nothing matches '%s'", *solvable);
-          continue;
-        }
-      pool_best_solvables (pool, &q, 0);
-      for (int j = 0; j < q.count; j++)
-        queue_push (&pile, q.elements[j]);
+      /* solvables prefixed by @ are file names */
+      if (**solvable == '@')
+        add_solvables_from_file_to_pile (*solvable + 1, pool, pile, error);
+      else
+        add_solvable_to_pile (*solvable, pool, pile);
+
+      if (*error)
+        return NULL;
     }
 
-  return pile;
+  return g_steal_pointer (&pile);
```

Comment:
does it really work? I always used `g_malloc0()` to create pointer.

If it really does and valgrind is happy - I'm fine.